### PR TITLE
Add the ability add/edit user comments on workflows [BW-759]

### PR DESCRIPTION
**Pull request by cahrens**
This PR allows the ability to specify a comment when launching a workflow, and the ability to view/edit the comment from either the Job History page or the Submission Details.

It introduces a reusable "ValidatedTextArea", based on our style guide. The existing "ValidatedInput" is for single-line input only.

It also does a bit of clean-up with regard to marking div's that display error messages as aria-live, and fixing some syntax that was throwing deprecation warnings.

See also https://github.com/broadinstitute/rawls/pull/1510

### CODE REVIEW

File: src/components/input.js
Comment (cahrens) [inline src/components/input.js]: This was an existing line of code from LaunchAnalysisModal (I take no credit for it…).

File: src/pages/workspaces/workspace/JobHistory.js
Comment (cahrens) [inline src/pages/workspaces/workspace/JobHistory.js:168]: @breilly2 I added the ability to specify an ariaLabel so that "No. of Workspaces" reads out correctly.

File: src/pages/workspaces/workspace/JobHistory.js
Comment (cahrens) [inline src/pages/workspaces/workspace/JobHistory.js]: @breilly2 This code was throwing deprecation warnings.

File: src/pages/workspaces/workspace/JobHistory.js
Comment (cahrens) [reply src/pages/workspaces/workspace/JobHistory.js:168]: I also figured that for the first column, "click for details" was not helpful (and potentially confusing) in the spoken text.

File: src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js
Comment (cahrens) [inline src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js:68]: This would show a pretty ugly error before, so pulling out just the message portion. Before:

image

File: src/style.css
Comment (cahrens) [inline src/style.css:91]: See the block of CSS immediately preceding this.

File: src/pages/workspaces/workspace/JobHistory.js
Comment (breilly2) [reply src/pages/workspaces/workspace/JobHistory.js]: Cool. Also, this could actually be `Utils.switchCase`, as Isaac pointed out to me in a similar situation in the showcase PR.

**Comment by kyuksel** [comment]
Pending scheduling mobbing time for review with @cahrens per policy.

**Review by aednichols — CHANGES_REQUESTED** [review]
First preference:

- Treat all unicode characters, including emoji, correctly & equally. We don't live in an ASCII world anymore 😃 

Second preference:

- User friendly error that does not quote raw SQL errors

---

Screenshot 2021-08-30 at 17-38-45 anichols-test - Workflows - echo_strings

**Comment by dmohs** [comment]
Nice catch @aednichols. The returned string is truncated (e.g., "Incorrect string value: '\\xF0\\x9F\\x98\\x80\\xF0\\x9F...' for column 'USER_COMMENT' at row 1"), so I don't think we can reliably parse it to display the right thing. The request payload (as rendered by Chrome) looks good to me, so I'd look to Rawls as the source of the bug.

I judge the details in the error message are useful, so I don't think I'd change this code. It seems to me it is doing the best it can with what it gets.

**Comment by cahrens** [comment]
Yes, thanks @aednichols for thinking to test with non-ASCII characters… my mind still defaults to an ASCII world.

It's interesting because some unicode characters, like ăѣ, work fine (and @salonishah11 included a unit test for them in https://github.com/broadinstitute/rawls/pull/1489), but others like 𝔠 or 😍, do not. As @dmohs commented, I think the terra-ui code is doing the best it can, unless we wanted to add client-side validation to exclude certain classes of characters. My thought is that we should probably create a new ticket for further investigation/changes (most likely on the rawls side). But I will continue to poke a bit more as well.

**Comment by aednichols** [comment]
The more common/"normal" non-ASCII characters used by major languages have lower indexes in the unicode numbering system, causing them to require only one byte in a possible multi-byte encoding. So that explains why some non-ASCII works but not all.

I tried some Polish characters (my other language) and they work fine as expected.

Screen Shot 2021-08-31 at 2 45 16 PM

**Comment by cahrens** [comment]
Right, I think the 4-byte characters are the problem… though supposedly mysql has the ability to support them. https://dev.mysql.com/doc/refman/5.6/en/charset-unicode.html

**Comment by aednichols** [comment]
I could look into the MySQL encoding configuration with you if you like, I find text encoding problems fascinating.

On the other hand, to save time it makes me feel better knowing that it works just fine with most characters/languages people are actually going to use.

**Comment by cahrens** [comment]
@aednichols I agree that character encoding is interesting, but I think we probably shouldn't get too distracted from the other things we are working on (and probably product should weigh in on how much they do or don't care). I'll create a separate ticket to capture this issue, and we can discuss with product whether it's worth spending time on it.

https://broadworkbench.atlassian.net/browse/BW-797

**Comment by aednichols** [comment]
Agreed.

Even if we did figure out a fix, it ~would~ could be a potentially invasive DB reconfiguration that we would have to weight against the chance of introducing new, worse bugs.

**Review by aednichols — APPROVED** [review]
LGTM, with an optional suggestion:

I did some informal usability testing and the focus group (my boyfriend) preferred this version because it states explicitly that it's optional

Screenshot 2021-08-31 at 11-45-29 anichols-test - Workflows - echo_strings

File: src/components/input.js
Comment (cahrens) [inline src/components/input.js]:
```suggestion
      { className: error ? 'error-style' : 'focus-style', ...inputProps }
```

File: src/components/input.js
Comment (cahrens) [inline src/components/input.js]: 324-325 can be same line.

File: src/components/input.js
Comment (dmohs) [inline src/components/input.js]:
```suggestion
 * Adds a WBR element before each period which advises the browser that a line-break here is acceptable.
```

File: src/components/input.js
Comment (cahrens) [reply src/components/input.js]:
```suggestion
const PeriodWrapper = ({ children }) => {
  const child = Children.only(children)
  const wrappableText = _.flow(
    _.split(/(\.)/),
    _.flatMap(sub => sub === '.' ? [wbr(), '.'] : sub)
  )(child)
  return h(Fragment, wrappableText)
}
```

File: src/components/input.js
Comment (cahrens) [inline src/components/input.js]: 0.75

File: src/pages/workspaces/workspace/JobHistory.js
Comment (cahrens) [inline src/pages/workspaces/workspace/JobHistory.js]: updatingCommentId

File: src/pages/workspaces/workspace/JobHistory.js
Comment (cahrens) [inline src/pages/workspaces/workspace/JobHistory.js]: Can just be a div.

Comment (cahrens) [reply src/pages/workspaces/workspace/JobHistory.js]: Font size of 12

File: src/pages/workspaces/workspace/JobHistory.js
Comment (cahrens) [inline src/pages/workspaces/workspace/JobHistory.js]: May not need userComment &&

And don't need tooltip: userComment

File: src/pages/workspaces/workspace/jobHistory/UpdateUserCommentModal.js
Comment (kyuksel) [inline src/pages/workspaces/workspace/jobHistory/UpdateUserCommentModal.js]: Nit:
```suggestion
    `The comment can be at most ${maxCommentLength} characters.` : undefined
```

File: src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js
Comment (dmohs) [inline src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js]: Maybe?
```suggestion
    }, [h(Fragment, wrappableOnPeriods(launchError))])
```

**Review by dmohs — APPROVED** [review]
Approved with changes as suggested via Mob review session.

File: src/components/input.js
Comment (cahrens) [reply src/components/input.js]: @dmohs we actually decided to remove the function completely (https://broadinstitute.slack.com/archives/C01EHNUM73R/p1630528774002800)

File: src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js
Comment (cahrens) [reply src/pages/workspaces/workspace/workflows/LaunchAnalysisModal.js]: This got further simplified.

File: src/components/input.js
Comment (cahrens) [reply src/components/input.js]: We ended up removing this because it wasn't actually useful for textual messages (since HTML wraps on whitespace by default).

File: src/pages/workspaces/workspace/JobHistory.js
Comment (cahrens) [reply src/pages/workspaces/workspace/JobHistory.js]: @breilly2 @zarsky-broad I decided to stop being lazy and changed this to Utils.switchCase.

**Comment by cahrens** [comment]
FYI, I have pushed up all the changes from the mob review.

**Comment by cahrens** [comment]
@petesantos @kyuksel I am hoping to merge this by EOD today. Please let me know if you have any concerns about that-- I am not sure if I am expected to wait for both of you to approve, or if the 2 checkmarks I have are considered sufficient.

**Comment by kyuksel** [comment]
> I am hoping to merge this by EOD today. Please let me know if you have any concerns about that-- I am not sure if I am expected to wait for both of you to approve, or if the 2 checkmarks I have are considered sufficient.

@cahrens Two 👍🏼s are sufficient to merge (as long as at least one is from a "[terra devs](https://github.com/orgs/DataBiosphere/teams/terra-devs)" group member) so you're good to merge 😊  Thanks for making sure though!

**Pull request merged**
